src/compiler/crystal/loader/unix.cr

```diff
@@ -0,0 +1,118 @@
+# object file linking process. Most are caused by limitations of `libdl`.
+#
+# * Only dynamic libraries can be loaded. Static libraries and object files
+#   are unsupported.
+# * All libraries are loaded into the same namespace. That means libraries
+#   loaded in the compiler program itself may provide symbols to libraries
+#   loaded with `Crystal::Loader`. Symbols may be available without explicitly
+#   mentioning their libraries. It might be impossible to link against other
+#   verions of the libraries that the compiler is linked against.
+# * A fully statically linked compiler may help dealing with the previous
+#   issue. But using `libdl` in a non-dynamically loaded executable might cause
+#   other issues.
+
+class Crystal::Loader
+  alias Handle = Void*
+
+  # Parses linker arguments in the style of `ld`.
+  def self.parse(args : Array(String), *, search_paths : Array(String) = default_search_paths) : self
+    libnames = [] of String
+    file_paths = [] of String
+    OptionParser.parse(args) do |parser|
+      parser.on("-L DIRECTORY", "--library-path DIRECTORY", "Add DIRECTORY to library search path") do |directory|
+        search_paths << directory
+      end
+      parser.on("-l LIBNAME", "--library LIBNAME", "Search for library LIBNAME") do |libname|
+        libnames << libname
+      end
+      parser.on("-static", "Do not link against shared libraries") do
+        raise LoadError.new "static libraries are not supported by Crystal's runtime loader"
+      end
+      parser.unknown_args do |args, after_dash|
+        file_paths.concat args
+      end
+    end
+
+    self.new(search_paths, libnames, file_paths)
+  end
+
+  def find_symbol?(name : String) : Handle?
+    @handles.each do |handle|
+      address = LibC.dlsym(handle, name)
+      if address
+        return address
+      end
+    end
+  end
+
+  def load_file(path : String | ::Path) : Handle
+    load_file?(path) || raise LoadError.new String.new(LibC.dlerror)
+  end
+
+  private def open_library(path : String)
+    LibC.dlopen(path, LibC::RTLD_LAZY | LibC::RTLD_GLOBAL)
+  end
+
+  # Closes all libraries loaded with this loader instance.
+  #
+  # Libraries are only unloaded when there are no references left.
+  def close_all : Nil
+    @handles.each do |handle|
+      LibC.dlclose(handle)
+    end
+  end
+
+  # Returns a list of directors used as the default search paths
+  def self.default_search_paths : Array(String)
+    default_search_paths = [] of String
+
+    if env_library_path = ENV[{{ flag?(:darwin) ? "DYLD_LIBRARY_PATH" : "LD_LIBRARY_PATH" }}]?
+      # TODO: Expand tokens $ORIGIN, $LIB, $PLATFORM
+      default_search_paths.concat env_library_path.split(Process::PATH_DELIMITER, remove_empty: true)
+    end
+
+    {% if flag?(:linux) || flag?(:bsd) %}
+      read_ld_conf(default_search_paths)
+    {% end %}
+
+    {% if flag?(:darwin) %}
+      default_search_paths << "/usr/lib"
+      default_search_paths << "/usr/local/lib"
```

Comment:
Should we add the `/opt/homebrew/lib` path for ARM64 Darwin? Or do that as a patch on the Homebrew-side? (This would mirror the behavior of the FFI gem, for example)

Reply:
We should keep the internals exactly as in `ld` to avoid confusion between linking in interpreted and compiled mode. Any user-defined paths must be configured via DYLD_LIBRARY_PATH or CRYSTAL_LIBRARY_PATH.